### DISCUSSION TITLE
Api 控制

### DISCUSSION THREAD

**Discussion by Ada19960111** [XTLS/Xray-core#2974] [Q&A]
大家好，是这样的，我之前使用过trojan-go，他提供了grpc的api控制，同时支持配置tls双向验证来保证安全，最近我开始使用xray，我发现xray也支持api操作，但是对于双向验证的配置没有示例，请问我该如何配置入站参数，来启用api的双向验证呢？如果能提供示例的配置文件就太感谢了，官方文档中的教程没有启动安全验证。。。

**Comment by Ada19960111** [comment]
还有就是，我本来准备加电报群的，结果进去就是一个标号为几的碳原子为手性碳原子，当场圆寂，经过了5分钟的学习，了解了个大概，信心满满选了答案，然后就被封禁了，还联系不到管理员，让我10分钟倒数中悬着的心终于死了，可以的话，麻烦大佬给我解禁一下吧

**Reply by maoxikun** [reply]
https://xtls.github.io/Xray-docs-next/config/transport.html#tlsobject

**Reply by Ada19960111** [reply]
谢谢回复，我有添加这个配置，这里附上我的配置文件内容，但是我尝试链接时，被拒绝了，服务端也没有日志打印，证书使用的是自签名的证书，cn包括了域名
```
{
  "api": {
    "tag": "api",
    "services": [
      "HandlerService",
      "LoggerService",
      "StatsService"
    ]
  },
  "inbounds": [
    {
      "listen": "0.0.0.0",
      "port": 10085,
      "protocol": "dokodemo-door",
      "settings": {
        "address": "0.0.0.0"
      },
      "streamSettings": {
        "network": "grpc",
        "security": "tls",
        "tlsSettings":{
          "serverName": "xxx.com",
          "certificates": [
            {
              "usage": "encipherment",
              "certificateFile": "/server.crt",
              "keyFile": "/server.key"
            },
            {
              "usage": "verify",
              "certificateFile": "/ca.crt"
            }
          ]
        }
      },
      "tag": "api"
    }
  ],
  "outbounds": [
  ],
  "rules": [
    {
      "domainStrategy": "IPIfNonMatch",
      "domainMatcher": "hybrid",
      "rules": [
        {
          "inboundTag": [
            "api"
          ],
          "outboundTag": "api",
          "type": "field"
        }
      ]
    }
  ],
  "log": {
    "loglevel": "debug"
  }
}
```

**Reply by Ada19960111** [reply]
java代码返回 UNIMPLEMENTED: unknown service xray.app.proxyman.command.HandlerService (我测试了删除outbound方法)

**Reply by maoxikun** [reply]
你说的是api的ssl啊，我不是很清楚api是否可以使用ssl，你不用证书能调通api接口吗，api的ssl你或许可以试试服务端前置nginx这种（nginx开启ssl双向认证）转发请求到api接口

**Reply by Ada19960111** [reply]
是可以的，我不添加tls部分内容，就可以正常执行调用，服务端日志也有体现，我会去看看nginx双向验证，但是我更倾向于使用xray自身的配置来完成这个功能，如果引入更多的元件，维护会成为问题，谢谢回答

**Reply by maoxikun** [reply]
刚试了一下好像可以了，你把"network": "grpc"删掉，dokodemo-door协议类似端口转发，所以可能不需要network？但好像是单向的，客户端可以不使用证书加密，只验证服务端证书，不知道是不是我证书有问题

**Reply by Ada19960111** [reply]
好的，我晚些试试，使用nginx的方法我试了是可以的，非常感谢提供方案

**Reply by Ada19960111** [reply]
我尝试去掉network配置项，确实可以使用了，但是也如你所说的那样，他只是启动了tls服务端验证，并没有验证来自客户端的证书，所以很遗憾不能安全通信，这样启动的接口任意一个客户端都可以操作api接口，这或许是一个issues？

**Reply by maoxikun** [reply]
xray这个貌似只支持单向证书认证，服务端使用证书私钥加密，客户端使用证书验证（可以跳过证书验证），或许可以提一个issue看看

**Reply by chise0713** [reply]
https://xtls.github.io/config/transport.html#certificateobject
```
usage: "encipherment" | "verify" | "issue"
```
xray 有验证选项

**Comment by maoxikun** [comment]
> https://xtls.github.io/config/transport.html#certificateobject
> 
> ```
> usage: "encipherment" | "verify" | "issue"
> ```
> 
> xray 有验证选项

即使是配置了verify，客户端没传证书过去也能成功连接
<img width="683" alt="image" src="https://github.com/XTLS/Xray-core/assets/106460972/ef3d6048-acc0-49da-b9bd-e34210337ea2">
<img width="1282" alt="image" src="https://github.com/XTLS/Xray-core/assets/106460972/bd1e94a6-c0e2-48a9-9242-9af14b562426">

**Reply by chise0713** [reply]
那可能是这个 verify 只生效于发起连接的客户端。主动验证似乎是 mTLS，不属于正常 TLS 协议。

**Reply by maoxikun** [reply]
但是nginx那种就是双向认证的，这么来看xray的这种方式确实是不安全的，即使服务端设置了证书，不管什么客户端都能连接
